docs(modal,pagination,pickerbutton): site docs to storybook

jawinn opened the pull request:
## Description

This continues the migration work to move the documentation from the deprecated docs site to Storybook, covering the components Modal, Pagination, and Picker button. Variants shown on the docs site for these components have been added to Storybook as new stories if they did not already exist. No custom MDX files were needed for these components.

### Component-specific changes and notes:

- Modal
  - Updates docs text and sets Docs story to `inline: false` to avoid it being cut off horizontally
- Pagination
  - Makes sure pagination variants are included in the Storybook docs. Note: the "button" style variant is represented in VRTs but not included for users as it is using the deprecated Split button component. I wasn't sure about the best way to document this as deprecated. That variant has no unique CSS and is basically just a Split button with two classes added.
- Picker button
  - Creates docs only stories to cover what is on the docs site
  - Moves old migration notes to the changelog
  - Note on the "Custom icon only" variant from the docs site: this appears to be using a workflow icon but is using the class `spectrum-PickerButton--uiicononly` that gets rid of the padding. It's unclear whether this class should be named something else that applies to both icon sets, if `spectrum-PickerButton--icononly` should use the same CSS (this currently is applied in Storybook if a workflow icon is used), if the component really should only allow UI icons, or if the wrong class was used in the example. I've included both a UI icon and a workflow icon on the Docs story in order to show that these render differently (without `spectrum-PickerButton--uiicononly` there is extra padding on the left and right).

## How and where has this been tested?

Please tag yourself on the tests you've marked complete to confirm the tests have been run by someone other than the author.

### Validation steps

@marissahuysentruyt ✅✅✅
- [Modal](https://pr-2854--spectrum-css.netlify.app/preview/?path=/docs/components-modal--docs) - includes all necessary information compared to the [docs site](https://opensource.adobe.com/spectrum-css/modal.html)
- [Pagination](https://pr-2854--spectrum-css.netlify.app/preview/?path=/docs/components-pagination--docs) - includes all necessary information compared to the [docs site 1](https://opensource.adobe.com/spectrum-css/pagination-explicit.html) [docs site 2](https://opensource.adobe.com/spectrum-css/pagination-listing.html)
- [Picker button](https://pr-2854--spectrum-css.netlify.app/preview/?path=/docs/components-picker-button--docs) - includes all necessary information compared to the [docs site](https://opensource.adobe.com/spectrum-css/pickerbutton.html)

### Regression testing

Expected VRT changes: label added to quiet Picker button.

Validate:

1. The documentation pages for at least two other components are still loading, including:

- [x] The pages render correctly, are accessible, and are responsive.

2. If components have been modified, VRTs have been run on this branch:

- [x] VRTs have been run and looked at.
- [x] Any VRT changes have been accepted (by reviewer and/or PR author), or there are no changes.

## To-do list

- [x] I have read the [contribution guidelines](/.github/CONTRIBUTING.md).
- [x] I have updated relevant storybook stories and templates.
- [x] If my change impacts **documentation**, I have updated the documentation accordingly.
- [x] ✨ This pull request is ready to merge. ✨

rise-erpelding reviewed (commented):
Hey, this looks great! Thanks for explaining all of your decision-making with this one in the PR description, that was really helpful in reviewing this.

I left a couple of minor suggestions for clarity but that was it!

marissahuysentruyt commented on components/pickerbutton/CHANGELOG.md:
```diff
@@ -208,128 +254,110 @@ See [Conventional Commits](https://conventionalcommits.org) for commit guideline
 - fix(pickerbutton): fix icon size
 
-- chore(pickerbutton): manual version increase for beta release
+### Migration guide
+
+#### .spectrum-PickerButton-UIIcon class removed
+
+The `.spectrum-PickerButton-UIIcon` class no longer matches our naming convention. Both types of icons now use the `spectrum-PickerButton-icon` class
```
Is this bit of the migration guide here specifically because these changes were made in version 3.0.34? How did you know where to put this section, as well as the other section about Quiet variant?

jawinn replied:
I looked at the git blame for the addition of the docs and found which release they were added in (also this is under version 4.0).

marissahuysentruyt commented on components/pickerbutton/stories/pickerbutton.stories.js:
```diff
@@ -108,35 +114,177 @@ export default {
+};
+
+export const Open = Template.bind({});
+Open.tags = ["docs-only"];
```
What's the distinction between using `vrt-only` and `docs-only` tags? They both remove the story from the sidebar, but our `docs-only` tagged stories have the chromatic snapshot disabled. How should I be deciding which one to use? We don't technically need the `vrt-only` tag if we want it to be included in the chromatic snapshot, right? 

I'll have to look this up!

jawinn replied:
The tags are now changed to use the newer built-in Chromatic tags though (per Cassondra's suggestion).

marissahuysentruyt commented on components/pickerbutton/stories/pickerbutton.stories.js:
```diff
@@ -108,35 +114,177 @@ export default {
+			})}
+			${Template({
+				...args,
+				iconName: "ArrowDown100",
```
Should we use the "ChevronDown" icon here? Or is that just a workflow icon?

jawinn replied:
This part of the template is to show a custom UI icon (other than the default Chevron).

marissahuysentruyt reviewed (changes requested):
@jawinn This looks good! I did leave a few questions for you. There were a few instances where I just wanted to make sure we didn't overlook things (like linking to a different component, or wording, etc). 

The biggest thing I wanted to bring up was the binding for pagination. I don't think I've had many components in these branches that have the chromatic coverage these 3 seem to have, so I might be doing it wrong or expecting the incorrect things. Let me know!

castastrophe commented on .storybook/decorators/context.js:
```diff
@@ -134,6 +134,11 @@ export const withContextWrapper = makeDecorator({
 					container.style.background = staticColorSettings[staticKey].background;
 				}
+
+				// Demonstrate the "Modal" component with a different darker background, so its bounds are visible.
+				if (rootClass == "spectrum-Modal") {
+					container.style.background = "var(--spectrum-transparent-black-400)";
+				}
```
I am really not sure about including a setting in the decorator that's specific to only one component and likely something we only want in test mode.

jawinn replied:
This helps things for the user, not just for test mode. Before it was unclear where the boundaries were, especially when changing the variant control. Example:
- https://opensource.adobe.com/spectrum-css/preview/?path=/story/components-modal--default&args=variant:fullscreen vs 
-  https://pr-2854--spectrum-css.netlify.app/preview/?path=/story/components-modal--default&args=variant:fullscreen

I don't like the idea of a component specific exception either.
1. Should I remove it and just let it display as-is?
2. What about doing a `customStyles` with `--mod-modal-background-color: var(--spectrum-background-layer-2-color);` instead to give the modal a contrasting color?
3. Something else?

marissahuysentruyt replied:
I'll let Cassondra make the final decision here, but from the options you mentioned, maybe the `customStyles` is the way to go. Is that more in line with how we're using `customStyles` in other components?

jawinn replied:
I just wanted to note that the `customStyles` option would be changing the opposite of what's on the PR now (component foreground instead of story body background). Which may not be desirable.

jawinn replied:
I've dropped that commit and have addressed this by displaying the Underlay component beside the Modal, similar to Dialog. I've also added a `showUnderlay` control.

castastrophe commented on components/pagination/stories/pagination.stories.js:
```diff
@@ -57,3 +74,15 @@ export default {
+ * Pagination's explicit variant uses the text field component to represent the current page, and text with the total number of pages. Along with action buttons for previous and next navigation.
+ */
+export const Explicit = Template.bind({});
+Explicit.tags = ["docs-only"];
```
Can you update the tags to our new standards when you get a chance?

jawinn replied:
Sure. This is done for all of the tags in this PR.

castastrophe commented on components/pagination/stories/template.js:
```diff
@@ -117,7 +119,10 @@ export const PaginationGroup = (args, context) => {
 			"gap": "32px",
 		})}>
-			${Template(args, context)}
+			${Template({
+				...args,
+				variant: "listing",
+			}, context)}
```
Ideally this only shows when window.isChromatic() is true

jawinn replied:
Some of the stories like this one (and Action button for example) display a different template for Chromatic that covers more options. We were calling this a kind of "kitchen sink" approach and it was partly to limit the number of snapshots from multiple stories. And also to cover testing specific things that we wanted to make sure are in the VRT.

You're right that it'd be a good not to show the chromatic testing (testing preview) view on the Docs / autodocs pages! I put up a fix on this PR that also checks if the `context.viewMode !== "docs"`. I also went ahead and made a similar but global change as part of PR #2885 that will prevent Chromatic templates from appearing on the Docs pages for _all_ components.

marissahuysentruyt replied:
Oh great! Can't wait to check out that chromatic testing PR, then!

jawinn commented on components/pagination/stories/pagination.stories.js:
```diff
@@ -57,3 +74,15 @@ export default {
 
 export const Default = PaginationGroup.bind({});
```
Yeah, I was flip-flopping on how to do that as well, with still getting it to appear first in the autodocs order. I think I'll just rename as "Default (Listing)".

jawinn commented on components/pagination/stories/pagination.stories.js:
```diff
@@ -57,3 +74,15 @@ export default {
 Default.args = {};
+
+/**
+ * Pagination's explicit variant uses the text field component to represent the current page, and text with the total number of pages. Along with action buttons for previous and next navigation.
```
Thanks. I've updated this wording.

marissahuysentruyt commented on components/pagination/stories/pagination.stories.js:
```diff
@@ -56,4 +73,19 @@ export default {
+ * number of pages.
+ */
+export const Explicit = Template.bind({});
+Explicit.tags = ["!dev"];
```
Do we want to keep `Explicit` in the sidenav? The prod Storybook has docs, default and explicit all in the sidenav. I've been keeping whatever was already in the sidenav, but maybe I shouldn't be. 🤷‍♀️

marissahuysentruyt replied:
Same question in regards to the `!dev` stories for picker button. Do we or should we keep any of those in the sidenav since that's how it is on prod? If it's ok that some are removed, how should I be making that decision on my branches?

jawinn replied:
This was to go along with how we were condensing the stories previously, since the variants are already in the VRT for the default story in the Chromatic-only template.

I think it's worth us all having a conversation soon about the conventions, because I think there is the case to be made for keeping more stories in the sidebar, that may have their chromatic snapshots disabled, for user's sake. In this case I could see keeping them in the sidebar.

jawinn replied:
For now I've added Pagination's "Explicit" back the sidebar, in addition to fixing the "Rendered more hooks than during the previous render" error that was happening when changing the variant control. This one I think is good to keep separate because they're significantly different sets of components (that used to have their own docs site pages), rather than a typical variant or state.

marissahuysentruyt commented on components/pickerbutton/stories/pickerbutton.stories.js:
```diff
@@ -108,35 +114,178 @@ export default {
 	},
 };
 
+// Story specific templates
+const SizingTemplate = (args) => html`
+	<div
```
I think the preference is to get markup like this into the `template.js` file as an export (Cassondra left me some requests regarding that on #2793 & #2833 for things like the variant checkbox group template and the divider sizes templates.)

jawinn replied:
I've moved these into the `template.js`.

marissahuysentruyt reviewed (commented):
Just a few more thoughts and questions for you! The things I mentioned before are all taken care of- nice work, per usual!

marissahuysentruyt commented on components/pickerbutton/stories/pickerbutton.stories.js:
```diff
@@ -108,35 +111,75 @@ export default {
 
 export const Disabled = Template.bind({});
-Disabled.tags = ["vrt-only"];
+Disabled.tags = ["!dev"];
```
Did we want to include the same sidebar items as on prod? Storybook prod has default, disabled, express, quiet (with label), and with label all listed.

jawinn replied:
I don't think all of these need to be there anymore (especially Express). We've been moving away from having stories for every single state and combination of states, and condensing as much as possible into the Default. And handling the large number of possible states in VRTs and Chromatic-only templates. Unless we're changing our approach there. I think the team still needs to have that conversation about conventions for this reorganization.

marissahuysentruyt commented on components/pickerbutton/stories/pickerbutton.stories.js:
```diff
@@ -108,35 +111,75 @@ export default {
 Quiet.args = {
-	isQuiet: true
+	isQuiet: true,
+	label: "All",
```
Adding the label here would change the chromatic snapshot, correct? It matches the docs site, so I like that update, but does that mean we should run chromatic on this branch? To update the baselines? And I guess if that's the case, modal might need to be checked as well (depending on what we decide to do with seeing the modal better).

jawinn replied:
Correct. I have that noted as an expected VRT change, and VRT should be run to finalize the validation instructions. Thanks for calling that out! I've run the VRTs and will check on the differences.

marissahuysentruyt reviewed (approved):
Just a couple more questions for me. Nothing I left comments on would prevent this from being merged, so I'm approving 👍🎉 Nice work~

rise-erpelding commented on components/pagination/stories/pagination.stories.js:
```diff
@@ -1,9 +1,11 @@
 /**
  * The pagination component displays numbered buttons or an input field to allow for navigation.
+ * 
+ * The default listing/page variant uses buttons for each page number.  
```
This is kind of a nitpick at this point and I don't have a strong opinion, but how would you feel about putting this as a comment on the Default story rather than the component overview?

jawinn replied:
Thanks, I've moved that to the Default story.

rise-erpelding commented on components/pickerbutton/stories/pickerbutton.stories.js:
```diff
@@ -1,9 +1,10 @@
 
 /**
- * The picker button component is used as a dropdown trigger. See Combobox.
+ * The picker button component is used as a dropdown trigger within other components such as combobox.
```
Nitpick: we could link to combobox here.

jawinn replied:
I've added that link.

rise-erpelding reviewed (approved):
I left a couple of small, nitpick-y things that you could optionally change if you wanted to, but otherwise I think this is looking pretty good!

castastrophe commented on components/pagination/stories/template.js:
```diff
@@ -103,21 +103,32 @@ export const Template = ({
 		<div style=${styleMap({
-			"display": window.isChromatic() ? "none" : undefined,
+			"display": window.isChromatic() && context.viewMode !== "docs" ? "none" : undefined,
 		})}>
 			${Template(args, context)}
 		</div>
 		<div style=${styleMap({
-			"display": window.isChromatic() ? "flex" : "none",
+			"display": window.isChromatic() && context.viewMode !== "docs" ? "flex" : "none",
```
The `.storybook/decorators/testing-preview.js` decorator does include a viewMode check already anyway:

```
window.isChromatic = typeof isChromatic === "function" && isChromatic() === true ? isChromatic : () => testingPreview && viewMode !== "docs";
```

jawinn replied:
I've removed those `viewMode` checks now that they're covered in the testing preview decorator changes that merged into main.

marissahuysentruyt commented on components/modal/stories/template.js:
```diff
@@ -27,6 +29,9 @@ export const Template = ({
 	`;
 
 	return html`
+		${when(showUnderlay,
+			() => Underlay({ isOpen })
+		)}
```
For dialog, I had to turn off the modal for the testing preview, so I'm really curious to see how we handle the testing preview for the modal itself!

marissahuysentruyt commented on components/pagination/stories/pagination.stories.js:
```diff
@@ -70,3 +89,17 @@ WithForcedColors.parameters = {
+ * and action buttons for the previous and next navigation. It also displays text with the total
+ * number of pages.
+ */
+export const Explicit = Template.bind({});
```
We should add the tags here, right (`!dev`)? To get this story out of the sidebar?

jawinn replied:
I'm purposefully including this one on the sidebar. It's a very different type of pagination that uses different elements/markup. I think it makes sense in this case, especially since these two types of pagination have their own separate docs site pages.

marissahuysentruyt replied:
I didn't re-read your last response to me on this. sorry!

marissahuysentruyt commented on components/pagination/stories/template.js:
```diff
@@ -99,21 +99,32 @@ export const Template = ({
 	return html`
 		<div style=${styleMap({
-			"display": window.isChromatic() ? "none" : undefined,
+			"display": window.isChromatic() && context.viewMode !== "docs" ? "none" : undefined,
```
Depending on what gets merged first, and since we already have some of this Chromatic preview stuff in place, did we want to refactor this `PaginationGroup` to use `Variants()`? 

Looks like it's being addressed in[ #2912 ](https://github.com/adobe/spectrum-css/pull/2912/files#diff-ae92bc343a4abb638a6bfdc1168db6fcab504115107fb5f7891062510c4acce5), but we might want to keep that though in our back pocket if this PR looks like it'll get merged first. 🤷‍♀️ How would you handle it?

jawinn replied:
Thanks for pointing that out in that upcoming PR. I'll have to address the rebase conflicts when it merges.

marissahuysentruyt commented on components/pickerbutton/stories/pickerbutton.stories.js:
```diff
@@ -113,6 +115,7 @@ export const Default = Template.bind({});
 Default.args = {};
 
 export const WithLabel = Template.bind({});
+WithLabel.storyName = "With label";
```
Do we need the `!dev` tag here, too?

jawinn replied:
I had left it how it was, but I think that is a good idea. I've added that tag to remove it from the sidebar in the latest update.

marissahuysentruyt commented on components/pickerbutton/stories/pickerbutton.stories.js:
```diff
@@ -130,11 +133,50 @@ WithForcedColors.parameters = {
 };
 
 export const Quiet = Template.bind({});
 Quiet.tags = ["!autodocs", "!dev", "test"];
```
I think I asked this on the autodocs config PR, but why are these Quiet and Disabled stories removed from the docs page? They're on the docs site right now, but are we "saving" them for just VRT? Should we be doing that for other components as well?

jawinn replied:
I've added these to the docs page.

marissahuysentruyt reviewed (commented):
This is looking good! Just looked at it with some fresh eyes and the modes stuff in mind, so I found a few more items we could address if you wanted!

jawinn commented:
> Just a couple of really minor things that I don't feel strongly about and don't change the PR approval: In Modal and Pagination, we have some of these controls that are lowercased instead of capitalized, which doesn't really match the way the rest of the controls are named, how would you feel about adding a "name" property to these to adjust that? 

I've adjusted the `name` property on those so the capitalization is consistent. I also added a couple default value summaries and a missing control description.

castastrophe commented on components/pagination/stories/pagination.stories.js:
```diff
@@ -58,7 +79,11 @@ export default {
+ * The default listing/page variant uses buttons for each page number. 
+ */
 export const Default = PaginationGroup.bind({});
+Default.storyName = "Default (Listing)";
```
```suggestion
Default.storyName = "Default (listing)";
```

Is listing a name or a descriptor? I think it should probably be lowercase since it describes the variant.

castastrophe commented on components/pagination/stories/pagination.stories.js:
```diff
@@ -70,3 +95,18 @@ WithForcedColors.parameters = {
 };
+
+// ********* DOCS ONLY ********* //
+/**
```
Great addition, thanks for adding this

castastrophe commented on components/pagination/stories/template.js:
```diff
@@ -98,4 +98,12 @@ export const Template = ({
 		</nav>
 	`;
+
+	if (variant === "explicit") {
+		return explicitVariant;
+	}
+	else if (variant == "button") {
+		return buttonVariant;
+	}
+	return listingVariant;
```
Very succinct, nice restructuring

castastrophe commented on components/pickerbutton/stories/pickerbutton.stories.js:
```diff
@@ -1,12 +1,14 @@
+
```
Can probably remove the empty whitespace at the top

castastrophe commented on components/pickerbutton/stories/pickerbutton.stories.js:
```diff
@@ -1,12 +1,14 @@
 
 /**
- * The picker button component is used as a dropdown trigger. See Combobox.
+ * The picker button component is used as a dropdown trigger within other components such as [combobox](?path=/docs/components-combobox--docs).
```
Nice internal linking here

castastrophe commented on components/pickerbutton/stories/pickerbutton.stories.js:
```diff
@@ -82,7 +85,9 @@ export default {
 			name: "Position",
-			type: { name: "string" },
+			description:
+				"Which side the picker button is displayed on. This determines which corners are rounded.",
```
```suggestion
				"Denotes which side of a form field the button is displayed; this influences which corners are rounded.",
```

castastrophe replied:
Minor phrasing suggestion

jawinn replied:
Thanks. This is updated.

castastrophe commented on components/pickerbutton/stories/pickerbutton.stories.js:
```diff
@@ -142,6 +146,48 @@ Disabled.parameters = {
+/**
+ * The `spectrum-PickerButton--rounded` class increases the amount of rounding on the rounded corners.
+ */
+export const Rounded = Template.bind({});
```
Great additions here, thanks for taking care of this!

castastrophe commented on components/pickerbutton/stories/template.js:
```diff
@@ -70,3 +71,66 @@ export const Template = ({
+ * Displays the component with a custom icon (instead of the chevron UI icon).
+ * Two examples are shown; with a custom UI icon and a custom Workflow icon.
+ */
+export const CustomIconTemplate = (args) => html`
```
Ideally I'd like to move away from hardcoding layouts and headings in the templates for the docs pages but I think this is an edge case where it makes sense for now. I'd love to brainstorm if there are ways we can support these custom demos without a lot of custom grids and headings.

castastrophe reviewed (approved):
A few minor suggestions but approved overall
